### PR TITLE
Week 4 assignments

### PR DESCRIPTION
**Pull request by adnan-dedhawala**
### 4.1
- Created Promise function from scratch
- Refer file CustomPromise.js for the implementation
- Refer customPromise.test.js for test cases

### 4.2
Screenshot 2022-10-29 at 11 31 26 AM

- solved the question by using setPrototypeOf function
- The person is the parent of the Teacher. Hence teacher is able to access the properties of Person

### 4.3
Screenshot 2022-10-29 at 11 27 59 AM

- Created a function to display the Fibonacci series using iterators

### CODE REVIEW

**Review by arqam-pesto — CHANGES_REQUESTED** [review]
Rebase the PR on Week 3

File: Week-4/4.1/CustomPromise.js [hunk 1]
```diff
@@ -0,0 +1,127 @@
+const STATE = {
```
Comment (arqam-pesto) [inline Week-4/4.1/CustomPromise.js:1]: Consider using Enum

File: Week-4/4.3/index.js [hunk 2]
```diff
@@ -0,0 +1,34 @@
...
+    }
+
+    return {
+      value:
+        this.count === 1
+          ? this.fiboFirst
+          : this.count === 2
+          ? this.fiboSecond
+          : nextFiboNumber,
+      done: !(this.max >= this.count)
```
Comment (arqam-pesto) [inline Week-4/4.3/index.js]: use conditional returns, than the ternary operators

**Review by arqam-pesto — CHANGES_REQUESTED** [review]

File: Week-4/4.3/index.js (hunk 2 above)
Comment (arqam-pesto) [reply Week-4/4.3/index.js]: or in this case a switch

File: Week-4/4.1/CustomPromise.js (hunk 1 above)
Comment (adnan-dedhawala) [reply Week-4/4.1/CustomPromise.js:1]: enums can be used only with typescript. In my case, I have implemented an enum by create an object.

Comment (adnan-dedhawala) [reply Week-4/4.1/CustomPromise.js:1]: https://www.sohamkamani.com/javascript/enums/

**Review by arqam-pesto — APPROVED** [review]

**Pull request merged**